drivers/flash/nrf_qspi_nor.c

```diff
@@ -503,26 +503,87 @@ static inline int qspi_nor_read_id(struct device *dev,
 	return 0;
 }
 
-static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
-			 size_t size)
+static inline nrfx_err_t read_non_aligned(struct device *dev, off_t addr,
+					  void *dest, size_t size)
 {
-	void *dptr = dest;
-	size_t dlen = size;
 	uint8_t __aligned(4) buf[4];
+	uint8_t *dptr = dest;
 
-	if (!dest) {
-		return -EINVAL;
+	off_t flash_offset = addr % 4;
+	off_t flash_prefix = (4 - flash_offset);
+
+	if (flash_prefix > size) {
+		flash_prefix = size;
 	}
 
-	/* read size must be non-zero multiple of 4 bytes */
-	if ((size > 0) && (size < 4U)) {
-		dest = buf;
-		size = sizeof(buf);
-	} else if (((size % 4U) != 0) || (size == 0)) {
+	off_t dest_prefix = (4 - ((off_t)dptr % 4)) % 4;
+
+	if (dest_prefix > size) {
+		dest_prefix = size;
+	}
+
+	off_t flash_suffix = (size - flash_prefix) % 4;
+	off_t flash_middle = size - flash_prefix - flash_suffix;
+
+	off_t shift = flash_prefix - dest_prefix;
+	nrfx_err_t res = NRFX_SUCCESS;
+
+	/* read from aligned flash to aligned memory */
+	if (flash_middle != 0) {
+		res = nrfx_qspi_read(dptr + dest_prefix, flash_middle,
+				     addr + flash_prefix);
+		qspi_wait_for_completion(dev, res);
+		if (res != NRFX_SUCCESS) {
+			return res;
+		}
+	}
+
+	/* perform shift in RAM */
+	if (shift > 0) {
+		for (off_t i = dest_prefix + flash_middle - 1;
+			   i >= dest_prefix; i--) {
+			dptr[i + shift] = dptr[i];
+		}
+	} else if (shift < 0) {
+		for (off_t i = flash_prefix;
+			   i < flash_prefix + flash_middle; i++) {
+			dptr[i] = dptr[i - shift];
+		}
+	}
+
+	/* read prefix */
+	if (flash_prefix != 0) {
+		res = nrfx_qspi_read(buf, 4, addr - flash_offset);
+		qspi_wait_for_completion(dev, res);
+		if (res != NRFX_SUCCESS) {
+			return res;
+		}
+		memcpy(dptr, buf + flash_offset, flash_prefix);
+	}
+
+	/* read suffix */
+	if (flash_suffix != 0) {
+		res = nrfx_qspi_read(buf, 4,
+				     addr + flash_prefix + flash_middle);
+		qspi_wait_for_completion(dev, res);
+		if (res != NRFX_SUCCESS) {
+			return res;
+		}
+		memcpy(dptr + flash_prefix + flash_middle, buf, flash_suffix);
+	}
+
+	return res;
+}
+
+static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
+			 size_t size)
+{
+	if (!dest) {
 		return -EINVAL;
 	}
-	/* address must be 4-byte aligned */
-	if ((addr % 4U) != 0) {
+
+	/* read size must be non-zero */
+	if (size == 0) {
 		return -EINVAL;
```

Comment:
The soc_flash_nrf and stm32 drivers return 0, when requested read len is 0. Can you check the other drivers? This should probably return 0 here.

Reply:
You are right. STM32 and other drivers return 0 in this case.